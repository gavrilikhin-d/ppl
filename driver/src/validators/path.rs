use std::path::PathBuf;

use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

/// Diagnostic for a wrong path
#[derive(Error, Debug, Diagnostic)]
#[error("path doesn't exist")]
#[diagnostic(code(driver::wrong_path))]
pub struct WrongPath {
    /// Span to the path
    #[label("path doesn't exist")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_path`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingPathError {
    #[error(transparent)]
    #[diagnostic(code(driver::io_error))]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    WrongPath(#[from] WrongPath),
}

/// Check that path exists
pub fn exists(path: &str) -> Result<PathBuf, ExistingPathError> {
    let path = PathBuf::from(path);
    if path.try_exists()? {
        return Ok(path);
    }
    Err(WrongPath { at: (0..0).into() }.into())
}

/// Diagnostic for a path, that's not a directory
#[derive(Error, Debug, Diagnostic)]
#[error("expected path to directory")]
#[diagnostic(code(driver::expected_directory))]
pub struct ExpectedDirectory {
    /// Span to the path
    #[label("isn't a directory")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_dir`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingDirError {
    #[error(transparent)]
    ExistingPathError(#[from] ExistingPathError),

    #[error(transparent)]
    ExpectedDirectory(#[from] ExpectedDirectory),
}

/// Check that path exists and is a directory
pub fn is_dir(path: &str) -> Result<PathBuf, ExistingDirError> {
    let p = exists(path)?;
    if p.is_dir() {
        return Ok(p);
    }
    Err(ExpectedDirectory { at: (0, 0).into() }.into())
}
