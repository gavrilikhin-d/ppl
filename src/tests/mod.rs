use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    clone,
    common_functions,
    constraints,
    constraints_in_constructor,
    consume_greater,
    deps,
    deref_member_ref,
    destructor,
    empty_block,
    empty_constructor,
    escaped_id,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    invalid_indentation,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multifile,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    reference_to_none,
    references,
    specify_variable_ty,
    star,
    string,
    supertraits,
    traits,
    type_as_value,
    trait_with_ref,
    type_of,
    wrong_initializer_type
}

#[test]
fn ppl() {
    use std::path::Path;

    use insta::assert_snapshot;
    use tempdir::TempDir;

    // Compile-time check that file exists
    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl/src/lib.ppl"));

    let temp_dir = TempDir::new("ppl").unwrap();
    let tmp = temp_dir.path();
    let name = "ppl";
    let dir = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl"));

    let res = crate::e2e::internal::compile(&tmp, dir);
    if let Err(err) = res {
        assert_snapshot!("ppl.error", err);
        return;
    }

    let hir = crate::e2e::internal::hir(&tmp, name, &dir);
    assert_snapshot!("ppl.hir", hir);

    let ir = crate::e2e::internal::ir(&tmp, name, &dir);
    assert_snapshot!("ppl.ir", ir);
}
