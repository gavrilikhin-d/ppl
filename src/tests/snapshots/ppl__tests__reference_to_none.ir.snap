---
source: src/tests/mod.rs
expression: ir
---
; ModuleID = 'main'
source_filename = "src/main.ppl"

%"Type<String>" = type { ptr }
%"Type<String>Impl" = type { %String, %Integer }
%String = type { ptr }
%Integer = type { ptr }

@"Type<String>" = private global %"Type<String>" zeroinitializer
@0 = private unnamed_addr constant [7 x i8] c"String\00", align 1
@1 = private unnamed_addr constant [4 x i8] c"foo\00", align 1

define private void @initialize() !dbg !3 {
  %1 = alloca %"Type<String>", align 8, !dbg !7
  %"Type<String>.data" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 0, !dbg !7
  %"Type<String>.name" = getelementptr inbounds %"Type<String>Impl", ptr %"Type<String>.data", i32 0, i32 0, !dbg !7
  %2 = call %String @string_from_c_string_and_length(ptr @0, i64 6), !dbg !8
  store %String %2, ptr %"Type<String>.name", align 8, !dbg !8
  %"Type<String>.size" = getelementptr inbounds %"Type<String>Impl", ptr %"Type<String>.data", i32 0, i32 1, !dbg !8
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !8
  store %Integer %3, ptr %"Type<String>.size", align 8, !dbg !8
  %4 = load %"Type<String>", ptr %1, align 8, !dbg !8
  store %"Type<String>" %4, ptr @"Type<String>", align 8, !dbg !8
  br label %return, !dbg !8

return:                                           ; preds = %0
  ret void
}

declare %String @string_from_c_string_and_length(ptr, i64)

declare %Integer @integer_from_i64(i64)

define void @main.execute() !dbg !9 {
  call void @initialize(), !dbg !10
  call void @initialize.1(), !dbg !11
  call void @"foo <:Reference<None>>"(), !dbg !12
  br label %return, !dbg !12

return:                                           ; preds = %0
  ret void
}

define void @"foo <:Reference<None>>"() !dbg !13 {
  %1 = call %String @string_from_c_string_and_length(ptr @1, i64 3), !dbg !14
  call void @"println <:String>"(%String %1), !dbg !14
  br label %return, !dbg !15

return:                                           ; preds = %0
  ret void
}

declare void @"println <:String>"(%String)

define private void @initialize.1() !dbg !16 {
  br label %return, !dbg !17

return:                                           ; preds = %0
  ret void
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "ppl", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, sysroot: "/")
!2 = !DIFile(filename: "src/main.ppl", directory: ".")
!3 = distinct !DISubprogram(name: "initialize", linkageName: "initialize", scope: !2, file: !2, line: 3, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!7 = !DILocation(line: 3, column: 5, scope: !3)
!8 = !DILocation(line: 0, scope: !3)
!9 = distinct !DISubprogram(name: "main.execute", linkageName: "main.execute", scope: !2, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!10 = !DILocation(line: 3, column: 5, scope: !9)
!11 = !DILocation(line: 2, column: 8, scope: !9)
!12 = !DILocation(line: 3, column: 4, scope: !9)
!13 = distinct !DISubprogram(name: "foo <:Reference<None>>", linkageName: "foo <:Reference<None>>", scope: !9, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!14 = !DILocation(line: 0, column: 27, scope: !13)
!15 = !DILocation(line: 0, column: 19, scope: !13)
!16 = distinct !DISubprogram(name: "initialize.1", linkageName: "initialize.1", scope: !9, file: !2, line: 2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!17 = !DILocation(line: 2, column: 8, scope: !16)
